lua/neogit/lib/finder.lua

```diff
@@ -0,0 +1,178 @@
+    if #picker:get_multi_selection() > 0 then
+      for _, item in ipairs(picker:get_multi_selection()) do
+        table.insert(selection, item[1])
+      end
+    else
+      table.insert(selection, action_state.get_selected_entry()[1])
+    end
+
+    if not selection[1] or selection[1] == "" then
+      return
+    end
+
+    actions.close(prompt_bufnr)
+
+    if allow_multi then
+      on_select(selection)
+    else
+      on_select(selection[1])
+    end
+  end
+
+  local actions = require("telescope.actions")
+  return function(_, map)
+    local commands = {
+      ["Select"] = select_action,
+      ["Close"] = function(...)
+        -- Make sure to notify the caller that we aborted to avoid hanging on the async task forever
+        on_select(nil)
+        actions.close(...)
+      end,
+      ["Next"] = actions.move_selection_next,
+      ["Previous"] = actions.move_selection_previous,
+      ["NOP"] = actions.nop,
+      ["MultiselectToggleNext"] = actions.toggle_selection + actions.move_selection_worse,
+      ["MultiselectTogglePrevious"] = actions.toggle_selection + actions.move_selection_better,
+    }
+
+    for mapping, command in pairs(config.values.mappings.finder) do
+      if command:match("^Multiselect") then
+        if allow_multi then
+          map({ "i" }, mapping, commands[command])
+        end
+      else
+        map({ "i" }, mapping, commands[command])
+      end
+    end
+
+    return false
+  end
+end
+
+---@return FinderOpts
+local function default_opts()
+  return {
+    layout_config = {
+      height = 0.3,
+      prompt_position = "top",
+      preview_cutoff = vim.fn.winwidth(0),
+    },
+    allow_multi = false,
+    border = false,
+    prompt_prefix = " > ",
+    previewer = false,
+    layout_strategy = "bottom_pane",
+    sorting_strategy = "ascending",
+    theme = "ivy",
+  }
+end
+
+---@class FinderOpts
+---@field layout_config table
+---@field allow_multi boolean
+---@field border boolean
+---@field prompt_prefix string
+---@field previewer boolean
+---@field layout_strategy string
+---@field sorting_strategy string
+---@field theme string
+
+---@class Finder
+---@field opts table
+---@field entries table
+---@field mappings function|nil
+local Finder = {}
+Finder.__index = Finder
+
+---@param opts FinderOpts
+---@return Finder
+function Finder:new(opts)
+  local this = {
+    opts = vim.tbl_deep_extend("keep", opts, default_opts()),
+    entries = {},
+    select_action = nil,
+  }
+
+  setmetatable(this, self)
+
+  return this
+end
+
+---Adds entries to internal table
+---@param entries table
+---@return Finder
+function Finder:add_entries(entries)
+  for _, entry in ipairs(entries) do
+    table.insert(self.entries, entry)
+  end
+  return self
+end
+
+---Engages finder and invokes `on_select` with the item or items, or nil if aborted
+---@param on_select fun(item: any|nil)
+function Finder:find(on_select)
+  if config.values.use_telescope then
+    local pickers = require("telescope.pickers")
```

Comment:
Same here, only import in the telescope enabled branch rather than the top of the file